README
= Ruby FileMagic Library Binding

== Summary

ruby-filemagic provides a ruby binding to GNU libmagic.


== Prerequisites

GNU file package to provide libmagic (tested with 5.04).  On
Debian-like Linux systems this is provided by package libmagic-dev.
For Solaris systems sunfreeware.com provides package SMCfile.


== Build and install

To build either:

    rake build

or run:

    ruby extconf.rb
    make

To install (non-gem) either:

    rake install

or run:

    make install


== Gem

To create a gem, build as above then either:

    rake gem

or run:

    gem build ruby-filemagic.gemspec


== Tests

Note that tests may fail if you have a different version of libmagic
since the file content descriptive text may vary slightly. For example,
a plain file gives these differing results:

  libmagic v5.04: text/plain; charset=us-ascii
  libmagic v4.26: text/plain charset=us-ascii

To run the unit tests:

    rake test

or run:

    ruby test/regress.rb


== Examples

    require 'filemagic'
    magic = FileMagic.new
    puts magic.file('/tmp/foo')
    magic.close


== Copying

The filemagic extension library is copywrited free software by Travis
Whitton <whitton@atlantic.net>. You can redistribute it under the terms
specified in the COPYING file of the Ruby distribution.


== WARRANTY

THIS SOFTWARE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



CHANGELOG
--------------------------------------------------------------------------------
2003/07/28 version 0.1.0

o Initial release (Travis Whitton <whitton@atlantic.net>)

--------------------------------------------------------------------------------
2003/07/30 version 0.1.1

o Added manual close method (Travis Whitton <whitton@atlantic.net>)
o Added unit test suite (Travis Whitton <whitton@atlantic.net>)

--------------------------------------------------------------------------------
2009/09/14 version 0.1.2

o Packaged gem (Ricardo Chimal, Jr. <ricardo@heroku.com>)
o Ruby 1.9 compatibility (http://github.com/red-agent)

--------------------------------------------------------------------------------
2010/09/04 version 0.1.3

o Fixes for ruby 1.9.2
o #flags and #flags= methods



test/regress.rb
require "test/unit"
require "filemagic"

class TestFileMagic < Test::Unit::TestCase
  def setup
    @oldwd = Dir.getwd
    Dir.chdir( File.dirname(__FILE__) )
  end

  def teardown
    Dir.chdir( @oldwd )
  end

  def test_file
    fm = FileMagic.new(FileMagic::MAGIC_NONE)
    res = fm.file("pyfile")
    assert_equal("a python script text executable", res)
    res = fm.file("pylink")
    assert_equal("symbolic link to `pyfile'", res)
    fm.close
    fm = FileMagic.new(FileMagic::MAGIC_SYMLINK)
    res = fm.file("pylink")
    assert_equal("a python script text executable", res)
    fm.close
    fm = FileMagic.new(FileMagic::MAGIC_SYMLINK | FileMagic::MAGIC_MIME)
    res = fm.file("pylink")
    assert_equal("text/plain; charset=us-ascii", res)
    fm.close
    fm = FileMagic.new(FileMagic::MAGIC_COMPRESS)
    res = fm.file("pyfile-compressed.gz")
    assert_equal('a python script text executable (gzip compressed data, was "pyfile-compressed", from Unix, last modified: Wed Jul 30 21:20:45 2003)', res)
    fm.close
  end

  def test_buffer
    fm = FileMagic.new(FileMagic::MAGIC_NONE)
    res = fm.buffer("#!/bin/sh\n")
    fm.close
    assert_equal("POSIX shell script text executable", res)
  end

  def test_check
    fm = FileMagic.new(FileMagic::MAGIC_NONE)
    res = fm.check("perl")
    fm.close
    assert_equal(res, 0)
  end

  def test_compile
    fm = FileMagic.new(FileMagic::MAGIC_NONE)
    res = fm.compile("perl")
    fm.close
    assert_equal(res, 0)
    File.unlink("perl.mgc")
  end

  def test_new
    assert_raise(ArgumentError) do
      FileMagic.new(FileMagic::MAGIC_COMPRESS, nil)
    end
    assert_nothing_raised do
      fm = FileMagic.new
      assert_equal(FileMagic::MAGIC_NONE, fm.flags)
    end
  end

  def test_flags
    fm = FileMagic.new(FileMagic::MAGIC_COMPRESS)
    assert_equal(FileMagic::MAGIC_COMPRESS, fm.flags)
    fm.flags = FileMagic::MAGIC_NONE
    assert_equal(FileMagic::MAGIC_NONE, fm.flags)
  end
end


test/leaktest.rb
require 'filemagic'

# watch -n 1 'ps aux | grep leaktest'

def do_file(filename)
  fm = FileMagic.new(0)
  file = fm.file(filename)
  fm.close
  return file
end

loop do
  puts do_file($0)
  sleep 1
end


filemagic.c
/*********************************************************

   filemagic.c

   a Ruby extension to bind Ruby to the libmagic library
   (ftp://ftp.astron.com/pub/file/)

   patterned very closely off of audiofile.c by jaredj which
   is patterned very very closely off of gdbm.c by matz ;-)

   $Author: twhitton $
   $Date: 2003/07/30 13:25:10 $

*********************************************************/

#include "ruby.h"
#include <magic.h>

#ifndef RSTRING_LEN
#define RSTRING_LEN(str) RSTRING(str)->len
#endif

static VALUE cFileMagic, rb_FileMagicError;

#define GetMagicCookie(obj, cookie) {\
    Data_Get_Struct(obj, struct magic_set, cookie);\
}

/* 
   GC never seems to happen until the program terminates, but this is called
   on any unclosed objects
*/
static void magick_free(magic_t cookie)
{
    magic_close(cookie);
}

/* This is mainly for subclassing */
static VALUE magick_init(VALUE self, VALUE flags)
{
    /* This isn't used internally, but might as well store it for children */
    rb_iv_set(self, "@flags", flags);
    return self;
}

/* Frees resources allocated */
static VALUE magick_close(VALUE class)
{
    magic_t cookie;
    GetMagicCookie(class, cookie);
    magic_close(cookie);
    /* This keeps magick_free from trying to free closed objects */
    RDATA(class)->data = NULL;
    return Qnil;
}

/* FileMagic.new */
static VALUE magick_new(int argc, VALUE *argv, VALUE obj)
{
    VALUE class = cFileMagic;
    VALUE flags;
    magic_t cookie;
    switch(argc) {
        case 0:
            flags = rb_const_get(class, rb_intern("MAGIC_NONE"));
            break;
        case 1:
            flags = argv[0];
            break;
        default:
            rb_raise(rb_eArgError, "wrong number of arguments (%i for 1)", argc);
            break;
    }
    cookie = magic_open(NUM2INT(flags));
    if (cookie == NULL) {
        rb_fatal("out of memory");
    }
    if (magic_load(cookie, NULL) == -1) {
        rb_fatal("%s", magic_error(cookie));
    }
    rb_FileMagicError = rb_define_class("FileMagicError", rb_eStandardError);
    obj = Data_Wrap_Struct(class, 0, magick_free, cookie);
    rb_obj_call_init(obj, 1, &flags);
    return obj;
}

/* Return a string describing file */
static VALUE magick_file(VALUE class, VALUE file)
{
    const char *m;
    magic_t cookie;
    m = StringValuePtr(file);
    GetMagicCookie(class, cookie);
    if ((m = magic_file(cookie, m)) == NULL)
        rb_raise(rb_FileMagicError, magic_error(cookie));
    return rb_str_new2(m);
}

/* Return a string describing the string buffer */
static VALUE magick_buffer(VALUE class, VALUE buffer)
{
    int i = RSTRING_LEN(buffer);
    const char *m;
    magic_t cookie;
    m = StringValuePtr(buffer);
    GetMagicCookie(class, cookie);
    if ((m = magic_buffer(cookie, m, i)) == NULL)
        rb_raise(rb_FileMagicError, magic_error(cookie));
    return rb_str_new2(m);
}

/* Checks validity of a magic database file */
static VALUE magick_check(VALUE class, VALUE file)
{
    int retval;
    const char *m;
    magic_t cookie;
    GetMagicCookie(class, cookie);
    m = StringValuePtr(file);
    retval = magic_check(cookie, m);
    return INT2FIX(retval);
}

/* Compiles a magic database file */
static VALUE magick_compile(VALUE class, VALUE file)
{
    int retval;
    const char *m;
    magic_t cookie;
    GetMagicCookie(class, cookie);
    m = StringValuePtr(file);
    retval = magic_compile(cookie, m);
    return INT2FIX(retval);
}

void Init_filemagic() {
    cFileMagic = rb_define_class("FileMagic", rb_cObject);
    rb_define_singleton_method(cFileMagic, "new", magick_new, -1);
    rb_define_method(cFileMagic, "initialize", magick_init, 1);
    rb_define_method(cFileMagic, "close", magick_close, 0);
    rb_define_method(cFileMagic, "file", magick_file, 1);
    rb_define_method(cFileMagic, "buffer", magick_buffer, 1);
    rb_define_method(cFileMagic, "check", magick_check, 1);
    rb_define_method(cFileMagic, "compile", magick_compile, 1);
    rb_define_attr(cFileMagic, "flags", 1, 1);
    rb_define_const(cFileMagic, "MAGIC_NONE", INT2FIX(MAGIC_NONE));
    rb_define_const(cFileMagic, "MAGIC_DEBUG", INT2FIX(MAGIC_DEBUG));
    rb_define_const(cFileMagic, "MAGIC_SYMLINK", INT2FIX(MAGIC_SYMLINK));
    rb_define_const(cFileMagic, "MAGIC_COMPRESS", INT2FIX(MAGIC_COMPRESS));
    rb_define_const(cFileMagic, "MAGIC_DEVICES", INT2FIX(MAGIC_DEVICES));
    rb_define_const(cFileMagic, "MAGIC_MIME", INT2FIX(MAGIC_MIME));
    rb_define_const(cFileMagic, "MAGIC_CONTINUE", INT2FIX(MAGIC_CONTINUE));
    rb_define_const(cFileMagic, "MAGIC_CHECK", INT2FIX(MAGIC_CHECK));
}
